src/widget/switch.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A toggle switch widget.
+
+use crate::kurbo::{Circle, Point, Rect, RoundedRect, Size};
+use crate::piet::{FontBuilder, Text, TextLayout, TextLayoutBuilder};
+use crate::piet::{LinearGradient, RenderContext, UnitPoint};
+use crate::theme;
+use crate::widget::Align;
+use crate::{
+    BaseState, BoxConstraints, Env, Event, EventCtx, LayoutCtx, PaintCtx, UpdateCtx, Widget,
+};
+
+#[derive(Debug, Clone)]
+pub struct Switch;
+
+impl Switch {
+    pub fn new() -> impl Widget<bool> {
+        Align::vertical(UnitPoint::CENTER, SwitchRaw::default())
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SwitchRaw {
+    knob_pos: Point,
+    knob_hovered: bool,
+}
+
+impl SwitchRaw {
+    fn knob_hit_test(&self, knob_width: f64, mouse_pos: Point) -> bool {
+        let knob_circle = Circle::new(self.knob_pos, knob_width / 2.);
+        if mouse_pos.distance(knob_circle.center) < knob_circle.radius {
+            return true;
+        }
+        false
+    }
+
+    fn paint_label(
+        &mut self,
+        paint_ctx: &mut PaintCtx,
+        base_state: &BaseState,
+        data: &bool,
+        env: &Env,
+        switch_width: f64,
+        switch_padding: f64,
+    ) {
+        let font_name = env.get(theme::FONT_NAME);
+        let font_size = env.get(theme::TEXT_SIZE_NORMAL);
+
+        let label = if *data { "ON" } else { "OFF" };
+
+        let font = paint_ctx
+            .text()
+            .new_font_by_name(font_name, font_size)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let text_layout = paint_ctx
+            .text()
+            .new_text_layout(&font, label)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let mut origin = UnitPoint::LEFT.resolve(Rect::from_origin_size(
+            Point::ORIGIN,
+            Size::new(
+                (base_state.size().width - text_layout.width()).max(0.0),
+                base_state.size().height + (font_size * 1.2) / 2.,
```

Comment:
Currently this widget gets messed up when squished below `BORDERED_WIDGET_HEIGHT`. That's a larger issue we need to deal with generally, but the quick fix here I believe would be to use `BORDERED_WIDGET_HEIGHT` instead of base_state height here.